scheduler/src/cook/mesos/data_locality.clj

```diff
@@ -1,48 +1,200 @@
+  (defn get-last-update-time
+    "Returns the last update time information"
+    []
+    @datasets->last-update-time-atom)
+
+  (defn jobs-to-update
+    "Return the list of job ids to update data locality costs for in the next iteration.
+     The number of jobs is limited by the configured batch size."
+    [db]
+    (let [{:keys [batch-size]} (config/data-local-fitness-config)
+          pending-jobs (->> db
+                            util/get-pending-job-ents
+                            (filter (fn [j] (not (empty? (:job/datasets j)))))
+                            (map (fn [{:keys [job/uuid job/submit-time] :as job}]
+                                   {:job/uuid uuid
+                                    :job/submit-time submit-time
+                                    :job/datasets (get-dataset-maps job)})))
+          datasets->last-update-time @datasets->last-update-time-atom
+          have-data?->job (group-by (fn [job] (contains? datasets->last-update-time (:job/datasets job)))
+                                    pending-jobs)
+          missing-data (->> (get have-data?->job false [])
+                            (sort-by :job/submit-time))
+          sorted-have-data (->> (get have-data?->job true [])
+                                (sort-by (fn [j] (datasets->last-update-time (:job/datasets j)))))
+          pending-job-datasets (->> pending-jobs
+                                    (map :job/datasets)
+                                    (into (hash-set)))
+          no-longer-waiting (set/difference (-> datasets->last-update-time keys set)
+                                            pending-job-datasets)]
+      {:to-fetch (into [] (take batch-size (concat missing-data sorted-have-data)))
+       :to-remove no-longer-waiting})))
+
+(defn fetch-data-local-costs
+  "Contacts the server to obtain the data local costs for the given job ids"
+  [jobs]
+  (let [{:keys [cost-endpoint]} (config/data-local-fitness-config)
+        job-uuid->datasets (into {} (map (fn [job] [(str (:job/uuid job)) (:job/datasets job)])
+                                         jobs))
+        request {:batch (UUID/randomUUID)
+                 :tasks (map (fn [job] {:task_id (str (:job/uuid job))
+                                        :datasets (:job/datasets job)})
+                             jobs)}
+        _ (log/debug "Updating data local costs :" (cheshire/generate-string request))
+        {:keys [body]} (http/post cost-endpoint {:body (cheshire/generate-string request)
+                                                 :content-type :json
+                                                 :accept :json
+                                                 :as :json-string-keys
+                                                 :spnego-auth true})
+        _ (log/debug "Got response:" body)
+        costs (body "costs")]
+    (->> costs
```

Comment:
Nitpick: we can inline `costs` here as it is used only once.

Reply:
Done